diff -Nur binutils-2.27-orig/libiberty/pex-unix.c binutils-2.27/libiberty/pex-unix.c
--- binutils-2.27-orig/libiberty/pex-unix.c	2016-08-03 09:36:54.000000000 +0200
+++ binutils-2.27/libiberty/pex-unix.c	2016-08-10 19:15:07.000000000 +0200
@@ -391,7 +391,7 @@
 
 /* Execute a child.  */
 
-#if defined(HAVE_SPAWNVE) && defined(HAVE_SPAWNVPE)
+#if defined(HAVE_SPAWNVE) && defined(HAVE_SPAWNVPE) && !defined(__QNX__)
 /* Implementation of pex->exec_child using the Cygwin spawn operation.  */
 
 /* Subroutine of pex_unix_exec_child.  Move OLD_FD to a new file descriptor
