--- binutils/libiberty/pex-unix.c-orig	2015-03-28 02:38:16.012000000 +0100
+++ binutils/libiberty/pex-unix.c	2015-03-28 02:38:21.956000000 +0100
@@ -392,7 +392,7 @@
 
 extern char **environ;
 
-#if defined(HAVE_SPAWNVE) && defined(HAVE_SPAWNVPE)
+#if defined(HAVE_SPAWNVE) && defined(HAVE_SPAWNVPE) && !defined(__QNX__)
 /* Implementation of pex->exec_child using the Cygwin spawn operation.  */
 
 /* Subroutine of pex_unix_exec_child.  Move OLD_FD to a new file descriptor
