diff --git a/lib/fixedsizehash.h b/lib/fixedsizehash.h
index dac2a5f5..4d46ee2f 100644
--- a/lib/fixedsizehash.h
+++ b/lib/fixedsizehash.h
@@ -30,7 +30,9 @@
 #include <errno.h>
 #include <stdlib.h>
 #include <string.h>
-#include <sys/cdefs.h>
+#if defined (__GLIBC__) || defined (__UCLIBC__)
+# include <sys/cdefs.h>
+#endif
 
 #include <system.h>
 
diff --git a/lib/system.h b/lib/system.h
index 172456d2..1701f1e4 100644
--- a/lib/system.h
+++ b/lib/system.h
@@ -30,7 +30,9 @@
 #define LIB_SYSTEM_H	1
 
 #include <errno.h>
-#include <error.h>
+#if defined (__GLIBC__) || defined (__UCLIBC__)
+# include <error.h>
+#endif
 #include <stddef.h>
 #include <stdint.h>
 #include <string.h>
