diff --git a/Makefile.am b/Makefile.am
index 9c47afa9..413c3d57 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -26,8 +26,11 @@ AM_MAKEFLAGS = --no-print-directory
 
 pkginclude_HEADERS = version.h
 
-SUBDIRS = config lib libelf libcpu backends libebl libdwelf libdwfl libdw \
-	  libasm debuginfod src po doc tests
+SUBDIRS = config lib libelf
+if LIBDW
+SUBDIRS += libcpu backends libebl libdwelf libdwfl libdw \
+	   libasm debuginfod src po doc tests
+endif
 
 EXTRA_DIST = elfutils.spec GPG-KEY NOTES CONTRIBUTING \
 	     COPYING COPYING-GPLV2 COPYING-LGPLV3
diff --git a/config/Makefile.am b/config/Makefile.am
index 0d3ba164..bb37ff87 100644
--- a/config/Makefile.am
+++ b/config/Makefile.am
@@ -33,7 +33,10 @@ EXTRA_DIST = elfutils.spec.in known-dwarf.awk 10-default-yama-scope.conf \
 	     debuginfod.service debuginfod.sysconfig profile.sh.in profile.csh.in
 
 pkgconfigdir = $(libdir)/pkgconfig
-pkgconfig_DATA = libelf.pc libdw.pc
+pkgconfig_DATA = libelf.pc
+if LIBDW
+pkgconfig_DATA += libdw.pc
+endif
 if LIBDEBUGINFOD
 pkgconfig_DATA += libdebuginfod.pc
 
diff --git a/configure.ac b/configure.ac
index 48071165..8acb66f5 100644
--- a/configure.ac
+++ b/configure.ac
@@ -747,6 +747,9 @@ AS_IF([test "x$enable_libdebuginfod" != "xno"], [
 AC_CHECK_LIB(pthread, pthread_setname_np, [
                       AC_DEFINE([HAVE_PTHREAD_SETNAME_NP],[1],[Enable pthread_setname_np])])
 
+AC_ARG_ENABLE([libdw], AC_HELP_STRING([--disable-libdw], [Disable libdw]))
+AM_CONDITIONAL(LIBDW, test "x$enable_libdw" != "xno")
+
 AS_IF([test "x$enable_libdebuginfod" = "xyes" || test "x$enable_libdebuginfod" = "xdummy"],
       [AC_DEFINE([ENABLE_LIBDEBUGINFOD], [1], [Enable libdebuginfod])])
 AS_IF([test "x$enable_libdebuginfod" = "xdummy"],
-- 
2.33.1

