diff -Nur elfutils-orig/lib/system.h elfutils/lib/system.h
--- elfutils-orig/lib/system.h	2019-02-15 16:46:47.000000000 +0100
+++ elfutils/lib/system.h	2019-04-05 11:54:14.000000000 +0200
@@ -33,6 +33,7 @@
 #include <error.h>
 #include <stddef.h>
 #include <stdint.h>
+#include <string.h>
 #include <sys/param.h>
 #include <endian.h>
 #include <byteswap.h>
diff -Nur elfutils-orig/libelf/elf_getaroff.c elfutils/libelf/elf_getaroff.c
--- elfutils-orig/libelf/elf_getaroff.c	2019-02-15 16:46:47.000000000 +0100
+++ elfutils/libelf/elf_getaroff.c	2019-04-05 11:54:14.000000000 +0200
@@ -38,7 +38,7 @@
 #include "libelfP.h"
 
 
-off_t
+int64_t
 elf_getaroff (Elf *elf)
 {
   /* Be gratious, the specs demand it.  */
diff -Nur elfutils-orig/libelf/elf_getbase.c elfutils/libelf/elf_getbase.c
--- elfutils-orig/libelf/elf_getbase.c	2019-02-15 16:46:47.000000000 +0100
+++ elfutils/libelf/elf_getbase.c	2019-04-05 11:54:14.000000000 +0200
@@ -37,7 +37,7 @@
 #include "libelfP.h"
 
 
-off_t
+int64_t
 elf_getbase (Elf *elf)
 {
   return elf == NULL ? (off_t) -1 : elf->start_offset;
diff -Nur elfutils-orig/libelf/elf_getdata_rawchunk.c elfutils/libelf/elf_getdata_rawchunk.c
--- elfutils-orig/libelf/elf_getdata_rawchunk.c	2019-02-15 16:46:47.000000000 +0100
+++ elfutils/libelf/elf_getdata_rawchunk.c	2019-04-05 11:54:14.000000000 +0200
@@ -41,7 +41,7 @@
 #include "common.h"
 
 Elf_Data *
-elf_getdata_rawchunk (Elf *elf, off_t offset, size_t size, Elf_Type type)
+elf_getdata_rawchunk (Elf *elf, int64_t offset, size_t size, Elf_Type type)
 {
   if (unlikely (elf == NULL))
     return NULL;
diff -Nur elfutils-orig/libelf/elf_update.c elfutils/libelf/elf_update.c
--- elfutils-orig/libelf/elf_update.c	2019-02-15 16:46:47.000000000 +0100
+++ elfutils/libelf/elf_update.c	2019-04-05 11:55:52.000000000 +0200
@@ -91,6 +91,7 @@
 	 fallback case might fail (with unexpected errnos) in some cases.
 	 So we only report an error when the call fails and errno is
 	 ENOSPC. Otherwise we ignore the error and treat it as just hint.  */
+#if 0
       if (elf->parent == NULL
 	  && (elf->maximum_size == ~((size_t) 0)
 	      || (size_t) size > elf->maximum_size))
@@ -116,6 +117,7 @@
 	    }
 
 	}
+#endif
 
       /* The file is mmaped.  */
       if ((class == ELFCLASS32
@@ -164,7 +166,7 @@
 }
 
 
-off_t
+int64_t
 elf_update (Elf *elf, Elf_Cmd cmd)
 {
   size_t shnum;
diff -Nur elfutils-orig/libelf/libelfP.h elfutils/libelf/libelfP.h
--- elfutils-orig/libelf/libelfP.h	2019-02-15 16:46:47.000000000 +0100
+++ elfutils/libelf/libelfP.h	2019-04-05 11:54:14.000000000 +0200
@@ -40,7 +40,7 @@
 #include <string.h>
 
 /* gettext helper macros.  */
-#define _(Str) dgettext ("elfutils", Str)
+#define _(Str) Str
 
 
 /* Helper Macros to write 32 bit and 64 bit functions.  */
diff -Nur elfutils-orig/libelf/libintl.h elfutils/libelf/libintl.h
--- elfutils-orig/libelf/libintl.h	1970-01-01 01:00:00.000000000 +0100
+++ elfutils/libelf/libintl.h	2019-04-05 11:54:14.000000000 +0200
@@ -0,0 +1 @@
+/* Empty stub for Android. */
diff --git a/configure.ac b/configure.ac
index b4e012de..28d4c2b8 100644
--- a/configure.ac
+++ b/configure.ac
@@ -485,7 +485,7 @@ if test "$libc_has_argp" = "false" ; then
 	AC_CHECK_LIB([argp], [argp_parse], [have_argp="true"], [have_argp="false"])
 
 	if test "$have_argp" = "false"; then
-		AC_MSG_ERROR("no libargp found")
+		AC_MSG_WARN("no libargp found")
 	else
 		argp_LDADD="-largp"
 	fi
