diff -Nur m4-orig/lib/sigsegv.c m4/lib/sigsegv.c
--- m4-orig/lib/sigsegv.c	2021-05-19 16:36:49.000000000 +0200
+++ m4/lib/sigsegv.c	2021-11-06 20:31:46.000000000 +0100
@@ -578,7 +578,10 @@
      - 'ucontext_t' and 'struct __darwin_ucontext' in <sys/_types/_ucontext.h>,
      - 'struct __darwin_mcontext64' in <arm/_mcontext.h>, and
      - 'struct __darwin_arm_thread_state64' in <mach/arm/_structs.h>.  */
-#  define SIGSEGV_FAULT_STACKPOINTER  ((ucontext_t *) ucp)->uc_mcontext->__ss.__sp
+#  if !defined __darwin_arm_thread_state64_get_sp
+#   define __darwin_arm_thread_state64_get_sp(ts) ((ts).__sp)
+#  endif
+#  define SIGSEGV_FAULT_STACKPOINTER  __darwin_arm_thread_state64_get_sp (((ucontext_t *) ucp)->uc_mcontext->__ss)
 
 # elif defined __powerpc__
 
