diff --git a/src/subprocess-posix.cc b/src/subprocess-posix.cc
index 8e78540..a19846b 100644
--- a/src/subprocess-posix.cc
+++ b/src/subprocess-posix.cc
@@ -36,6 +36,33 @@ extern char** environ;
 
 using namespace std;
 
+#if defined(USE_PPOLL) && defined(__linux__)
+#include <sys/syscall.h>
+#ifndef SYS_ppoll
+
+#define ppoll PPollFallback
+
+#ifdef __arm__
+#define SYS_ppoll (__NR_SYSCALL_BASE + 336)
+#else
+#error Please implement for your architecture
+#endif
+
+static int PPollFallback(struct pollfd* fds, nfds_t nfds,
+                         const struct timespec* timeout,
+                         const sigset_t* sigmask) {
+  /* Based on the implementation in glibc: */
+  struct timespec tv;
+  if (timeout != NULL) {
+    tv = *timeout;
+    timeout = &tv;
+  }
+  return syscall(SYS_ppoll, fds, nfds, timeout, sigmask, _NSIG / 8);
+}
+
+#endif
+#endif
+
 Subprocess::Subprocess(bool use_console) : fd_(-1), pid_(-1),
                                            use_console_(use_console) {
 }
