diff -Nur openssl-orig/Configurations/windows-makefile.tmpl openssl/Configurations/windows-makefile.tmpl
--- openssl-orig/Configurations/windows-makefile.tmpl	2021-03-25 14:28:38.000000000 +0100
+++ openssl/Configurations/windows-makefile.tmpl	2021-06-04 20:40:36.250952900 +0200
@@ -359,7 +359,7 @@
 
 libclean:
 	"$(PERL)" -e "map { m/(.*)\.dll$$/; unlink glob """{.,apps,test,fuzz}/$$1.*"""; } @ARGV" $(SHLIBS)
-	-del /Q /F $(LIBS) libcrypto.* libssl.* ossl_static.pdb
+	-del /Q /F $(LIBS) libcrypto.* libssl.*
 
 clean: libclean
 	{- join("\n\t", map { "-del /Q /F $_" } @PROGRAMS) -}
@@ -424,8 +424,6 @@
 				       "$(INSTALLTOP)\include\openssl"
 	@"$(PERL)" "$(SRCDIR)\util\mkdir-p.pl" "$(libdir)"
 	@"$(PERL)" "$(SRCDIR)\util\copy.pl" $(INSTALL_LIBS) "$(libdir)"
-	@if "$(SHLIBS)"=="" \
-	 "$(PERL)" "$(SRCDIR)\util\copy.pl" ossl_static.pdb "$(libdir)"
 
 uninstall_dev:
 
